src/Common/Core/Impl/Services/ServiceManager.cs

```diff
@@ -0,0 +1,149 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License. See LICENSE in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Reflection;
+using Microsoft.Common.Core.Diagnostics;
+
+namespace Microsoft.Common.Core.Services {
+    public class ServiceManager : IServiceManager {
+        class ServiceFactory {
+            public IServiceFactory Factory { get; set; }
+            public object[] Parameters { get; set; }
+        }
+
+        private readonly Dictionary<Type, object> _services = new Dictionary<Type, object>();
+        private readonly Dictionary<Type, ServiceFactory> _deferredServices = new Dictionary<Type, ServiceFactory>();
```

Comment:
I though that different ctors can have different number of arguments. Seemed more difficult with Func